cdap-ui/app/directives/dag-plus/my-dag-ctrl.js

```diff
@@ -275,67 +274,91 @@ angular.module(PKG.name + '.commons')
     }
 
     function initNodes() {
-      vm.instance.unmakeEverySource();
-      vm.instance.unmakeEveryTarget();
-
-      // This is to handle special case when we open a pipeline with only one node,
-      // thought not sure why jsPlumb behaves differently in this case. Exclude
-      // condition nodes since we add endpoints differently for those
-      if ($scope.nodes.length === 1 && $scope.nodes[0].type !== 'condition') {
-        vm.instance.deleteEveryEndpoint();
-      }
-
       angular.forEach($scope.nodes, function (node) {
         if (node.type === 'condition') {
           initConditionNode(node.name);
         } else if (node.type === 'splittertransform') {
           initSplitterNode(node);
         } else {
-          initNormalNode(node.name);
+          initNormalNode(node);
         }
 
-        vm.instance.makeTarget(node.name, {
-          isTarget: true,
-          dropOptions: { hoverClass: 'drag-hover' },
-          anchor: 'ContinuousLeft',
-          allowLoopback: false
-        });
+        if (!vm.instance.isTarget(node.name)) {
+          let targetOptions = {
+            isTarget: true,
+            dropOptions: { hoverClass: 'drag-hover' },
+            anchor: 'ContinuousLeft',
+            allowLoopback: false
+          };
+          if (node.type === 'alertpublisher') {
+            targetOptions.scope = 'alertScope';
+          } else if (node.type === 'errortransform') {
+            targetOptions.scope = 'errorScope';
+          }
+          vm.instance.makeTarget(node.name, targetOptions);
+        }
       });
     }
 
-    function initNormalNode(nodeName) {
-      let sourceObj = {
-        isSource: true,
-        filter: function(event) {
-          // we need this variable because when the user clicks on the endpoint circle, multiple
-          // 'mousedown' events are fired
-          if (event.target.className.indexOf('endpoint-circle') !== -1 && !endpointClicked) {
-            endpointClicked = true;
-          }
+    function addEndpointForNormalNode(endpointId, customConfig, nodesWithEndpoints, nodeName) {
```

Comment:
Just a suggestion, if we have a function that will add endpoint for normal nodes then should we be having a function that adds endpoints for other types of nodes? We seem to be using `vm.instance.addEndpoint` across the controller.

Reply:
Added `addEndpointForConditionNode` and `addEndpointForSplitterNode`